teps/0059-skip-guarded-task-only.md

```diff
@@ -249,6 +287,263 @@ Users should be able to specify that a guarded `Task` only should be skipped whe
+
+### Simplicity
+
+By scoping the skipping strategy to `WhenExpressions` only, we provide the flexibility safely with a minimal change. We also limit the interleaving of `Pipeline` graph execution paths and maintain the simplicity of the workflows.
+
+### Flexibility
+
+This proposal gives users more control to specify what happens when each guarded `Task` is skipped.
+
+## Upgrade & Migration Strategy
+
+<!--
+Use this section to detail wether this feature needs an upgrade or
+migration strategy. This is especially useful when we modify a
+behavior or add a feature that may replace and deprecate a current one.
+-->
+
+To enable users to smoothly transition from the old syntax to the new syntax, we will initially support both sytanxes. Then we will deprecate the old syntax and support the new syntax only as we go towards v1 API.
+
+
+## Alternatives
+
+<!--
+What other approaches did you consider and why did you rule them out?  These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
+
+### Skipping Policies
+
+Add a field - `whenSkipped` - that can be set to `runBranch` to unblock or `skipBranch` to block the execution of `Tasks` that are dependent on the guarded `Task`.
+
+```go
+type SkippingPolicy string
+
+const (
+    RunBranch  SkippingPolicy = "runBranch"
+    SkipBranch SkippingPolicy = "skipBranch"
+)
+```
+
+```yaml
+tasks:
+- name: task
+  when:
+  - input: foo
+    operator: in
+    values: [ bar ]
+  whenSkipped: runBranch / skipBranch
+  taskRef:
+  - name: task
+```
+Another option would be a field - `whenScope` - than can be set to `Task` to unblock or `Branch` to block the execution of `Tasks` that are dependent on the guarded `Task`.
+```go
+type WhenScope string
+
+const (
+    Task   WhenScope = "task"
+    Branch WhenScope = "branch"
+)
+```
+
+```yaml
+tasks:
+- name: task
+  when:
+  - input: foo
+    operator: in
+    values: [ bar ]
+  whenScope: task / branch
+  taskRef:
+  - name: task
+```
+
+However, it won't be clear that the skipping policies are related to `WhenExpressions` specifically.
+
+### Execution Policies
+
+Add a field - `executionPolicies` - that takes a list of execution policies for the skipping and failure strategies for given `Task`.
+This would align well with [TEP-0050: Ignore Task Failures](https://github.com/tektoncd/community/blob/main/teps/0050-ignore-task-failures.md) and is easily extensible. 
+
+```go
+type ExecutionPolicy string
+
+const (
+    IgnoreFailure     ExecutionPolicy = "ignoreFailure"
+    ContinueAfterSkip ExecutionPolicy = "continueAfterSkip"
+    ...
+)
+```
+
+```yaml
+tasks:
+- name: task
+  when:
+  - input: foo
+    operator: in
+    values: [ bar ]
+  executionPolicies:
+  - ignoreFailure
+  - continueAfterSkip
+  taskRef:
+  - name: task
+```
+
+However, it won't be clear that the skipping policies are related to `WhenExpressions` specifically. 
```

Comment:
This alternative can be further refined to align well with TEP-0050. 

But, I dont see this as a con, in the entirety, `continue on error` be it a process failure or skipped due to `when` expressions.

Reply:
> This alternative can be further refined to align well with TEP-0050.

please share any ideas you have to refine this alternative 😀 

> But, I dont see this as a con, in the entirety, `continue on error` be it a process failure or skipped due to `when` expressions.

I don't think we'd want to refer to skipping of a task as an error (skipping doesn't cause to pipeline failure)

It'll be more extensible and scalable to pass in the individual execution policies as a list instead of providing execution policies that are combinations of other execution policies, what do you think?

Reply:
@vdemeester, @sbwsg, @bobcatfish I'd appreciate your feedback on this alternative as well

Reply:
I do like the locality of `scope` to `when`. Is there a risk here that we have collisions in the meaning given by `scope` and any meaning given by changes from TEP-0050?  For example, are we going to end up in a situation where `when.scope` could be `branch` but TEP-0050 has introduced a field that circumvents this `scope`?  And then I guess a follow-up question would be: does this potential conflict of meaning matter?

If we introduce multiple fields that can interact in complex ways like this then we should probably try and get a handle on what those interactions will be and decide if there are design ramifications we can deal with now, or if we'll disallow them entirely through validation, or simply document them, or ... something else.

I went over and re-read TEP-0050 again but because it's at proposal stage atm I'm not totally clear how the concrete implementation there will shake out wrt this feature.

Apologies if I've misunderstood the overlap here!

Reply:
> I do like the locality of `scope` to `when`. Is there a risk here that we have collisions in the meaning given by `scope` and any meaning given by changes from TEP-0050? For example, are we going to end up in a situation where `when.scope` could be `branch` but TEP-0050 has introduced a field that circumvents this `scope`? And then I guess a follow-up question would be: does this potential conflict of meaning matter?
> 

TEP-0050 is about ignoring `Task` failures only (does not include `Task` skipping) so I believe that it won't introduce a field that does the same thing as `scope` or circumvents `scope`

> If we introduce multiple fields that can interact in complex ways like this then we should probably try and get a handle on what those interactions will be and decide if there are design ramifications we can deal with now, or if we'll disallow them entirely through validation, or simply document them, or ... something else.
> 

If we go forward with `scope` under `when`, we won't need to add a `continueOnSkip` execution policy

Added this alternative because it provides an easily extensible way for users to specify `execution policies` to define workflow behavior for each `PipelineTask` at one point -- and it'll be well aligned with TEP-0050 which would only add `ignoreFailure` execution policy

Reply:
> I went over and re-read TEP-0050 again but because it's at proposal stage atm I'm not totally clear how the concrete implementation there will shake out wrt this feature.

Yup, its a long overdue, hoping to have the proposal in by end of this week 🤞

Reply:
Ahh OK I think I understand better now - it's not so much whether there's going to be a conflict in functionality but whether consolidating these two concepts under a single section (`executionPolicies`) would make sense.

`executionPolicies` seems to me like a more general statement on the intended behaviour of the PipelineTask. And here we're dealing very directly with the outcome of a conditional.  Putting this under `executionPolicies` would be kind of like having an `ifElsePolicies` in a programming language which changes the behaviour of an `if` statement depending on those policies.  From my pov conditionals should be a relatively known quantity - it would be much more difficult to reason about complex bits of code if that were configurable elsewhere.  We're not writing a programming language but I do wonder if a similar effect would happen here?

> If we go forward with scope under when, we won't need to add a continueOnSkip execution policy

The lingering question on my mind is whether the skips we're discussing here will always be tied to `when`.  Will there ever come a time when `continueOnSkip` makes sense for _other_ scenarios?  This is probably not something we can answer right now but I feel like if it does happen we can always introduce `continueOnSkip` with semantics that make sense at that time.  In other words, maybe one day we consolidate `when.scope` with a `continueOnSkip` policy, if it makes sense to do so.

All of this to say: I like how closely tied the `scope` is to the `expressions` and I'm uncertain whether a more general skip policy should apply to the outcome of `when`.

Reply:
@sbwsg you make strong points about the options for an if statement should be relatively known, so should not be defined elsewhere, will include that in the tep as well! (your post wasn't rambly, reading your thought process was helpful! 😁)